### ISSUE TITLE
Correlations: Support JSON Transform

### ISSUE THREAD

**Issue by JakeCooper** [grafana/grafana#77767]
**Why is this needed**:

Currently we send structure logs like this to Grafana

![CleanShot 2023-11-06 at 21 52 54@2x](https://github.com/grafana/grafana/assets/5499880/bac00c3c-c845-4334-8653-ba925b27fc78)

To clean this up, we use a log_format

![CleanShot 2023-11-06 at 21 53 15@2x](https://github.com/grafana/grafana/assets/5499880/cd59cf1c-a3f8-4e84-be58-4af3e9aecde0)

We'd like to Create a Correlation for the "trace_id" property (as parsed by the JSON) to add 1-Click trace linking to our logs

**What would you like to be added**:

This dropdown should have a JSON parser, like the Explore Menu

![CleanShot 2023-11-06 at 21 54 17@2x](https://github.com/grafana/grafana/assets/5499880/77744780-32d8-4403-913e-ff374a0e3f08)

**Who is this feature for?**

People who like clean logs AND traces

**Comment by ifrost** [comment]
Thanks @JakeCooper! Definitely something on our radar. Do you use flat structure in your JSON (key+value) or you may have nested objects in your JSON that may also need to be extracted to links?

**Comment by gelicia** [comment]
Yeah, I feel like the "Extract Fields" transformation in dashboards is something very close to what we would need to do to let everyone get what they want out of their JSON data. The regex transformation has a non-optional need for an expression, but for JSON we could have it be that if an expression is not defined, we non-recursively get all variables out of the first level of the JSON. If an expression is defined, we extract that value out.

Getting items recursively or trying to construct a full list of non-flat JSON seems like we could run into performance issues. Not sure if we would want to build out more complex support. 

As an aside @JakeCooper you can use regex with the expression of `"trace_id": "(\w*\d*)"` to get this data if you don't want to wait :)

**Comment by JakeCooper** [comment]
> Thanks @JakeCooper! Definitely something on our radar. Do you use flat structure in your JSON (key+value) or you may have nested objects in your JSON that may also need to be extracted to links?

We do have nested JSON. Don't think anything we'd need on Correlate on tho.

Flat would suffice

Also will try the above and circle back @gelicia !

**Comment by JakeCooper** [comment]
Unforutnately @gelicia that doesn't seem to work :(
![CleanShot 2023-11-07 at 14 28 23](https://github.com/grafana/grafana/assets/5499880/85809a7c-9cde-41b8-814b-8e8f
![CleanShot 2023-11-07 at 14 28 34](https://github.com/grafana/grafana/assets/5499880/16cec221-1daa-4da9-8c96-061b64bfbee8)
ac3dc161)

**Comment by gelicia** [comment]
@JakeCooper This isn't clear in this interface (we ended up building a whole new [editor](https://github.com/grafana/grafana/pull/73315) with a [new transformations interface](https://github.com/grafana/grafana/pull/75930) because this one was complicated) but I have a hunch... for "results field" try putting `Line` and for transformation field, you can leave that blank. Expression can stay as provided, then Map value should be `trace_id` .

Then for your target query, you should be able to use `${trace_id}` to query for a specific trace. 

If this doesn't work, in explore, can you run the first query again, and open Query Inspector > Go to the Data tab. You'll see your logs in a table format. Let me know what column heading the data that has the JSON is in. That should be your "results field."

**Comment by JakeCooper** [comment]
Oh I was looking for the Explore interface! Looks like it was introduced in 10.2.0 2 weeks ago, and I'm on 10.1.5. Will upgrade

Also REALLY close here, it's now showing up, but it doesn't actually link out. I assume something with the ${trace_id} value is slightly misaligned.

![CleanShot 2023-11-07 at 14 51 57](https://github.com/grafana/grafana/assets/5499880/79db14b2-303c-4606-9a1a-eaf96ddc1996)

![CleanShot 2023-11-07 at 14 52 13](https://github.com/grafana/grafana/assets/5499880/b93b80ca-e775-4585-996f-896da2d2130f)

![CleanShot 2023-11-07 at 14 52 28](https://github.com/grafana/grafana/assets/5499880/70e71f9d-8faf-48d9-af3b-1edd8ef40e1c)

**Comment by JakeCooper** [comment]
Upgraded but unfortunately the Correlation editor doesn't seem to pickup on the trace stuff. Will continue fiddling and sharing as I go

![CleanShot 2023-11-07 at 15 01 29](https://github.com/grafana/grafana/assets/5499880/fda21744-0006-400a-87f7-7fda67924a39)

**Comment by gelicia** [comment]
@JakeCooper Unfortunately the json parsing that gets you that field list at the top happens after the correlation links are built, so we can't correlate directly off those.

Also unfortunately everything I'm seeing with what you built in the old editor looks like it should work. It shouldn't make a difference but if you're interested, maybe the new editor will work better.

 With the new editor, try clicking "Correlate with Time" or any of the fields (this just chooses what the button shows up next to), which will split the view and bring up the helper. 

(My expression is a little different than yours bc I have logfmt data, not Json)
<img width="728" alt="Screenshot 2023-11-07 at 5 10 59 PM" src="https://github.com/grafana/grafana/assets/1533128/99097478-67f0-49d1-9f8b-86401130448a">
<img width="534" alt="Screenshot 2023-11-07 at 5 11 53 PM" src="https://github.com/grafana/grafana/assets/1533128/a0963bfa-5796-44c3-be09-e0a256b61467">
<img width="1001" alt="Screenshot 2023-11-07 at 5 12 33 PM" src="https://github.com/grafana/grafana/assets/1533128/d27c6e2d-5bd3-4b92-83ca-228a7dc2edd4">

Thank you for sticking with this!

**Comment by JakeCooper** [comment]
Deepest apologies but the Transform Editor doesn't pop up :S. I'm so sorry I must be doing something really incorrectly 🙈 

Here's what I see:

https://github.com/grafana/grafana/assets/5499880/ffaec116-e709-471a-b041-c0e8633ddebd

**Comment by gelicia** [comment]
@JakeCooper 🤦 I am so incredibly sorry, the transformation editor got released in 10.3 which isn't out yet and I forgot. I am going to take another look at the regex I gave you because I'm wondering if there's some escaping happening I didnt account for.

**Comment by gelicia** [comment]
@JakeCooper Okay, does `"trace_id":"(\w*\d*)"` work as a regex? Without the space? Looking at the json formatted code in the first pic has a space but I think it's just added to make it pretty

**Comment by JakeCooper** [comment]
Empty Still unfortunately :(

![CleanShot 2023-11-07 at 16 07 07](https://github.com/grafana/grafana/assets/5499880/f29e278d-023b-4f14-9cab-6e328acd0610)

![CleanShot 2023-11-07 at 16 07 12](https://github.com/grafana/grafana/assets/5499880/aa27a31d-21e5-493c-bade-79737afa9d99)

**Comment by ifrost** [comment]
I think when you use `line_format` the original `Line` field will be replaced with the new message 🤔  In that case your regex transformation should not work (`Field` setting is empty so it defaults to "results field" which is `Line`). What's interesting is that the link is showing up meaning `${trace_id}` variable was picked somehow (links are shown only if variables are populated). Maybe you have a derived field called `trace_id`?

Anyway, one more thing you can try is to use `labels` field in you regex transform. From what I see in the details it seems like trace_id might be in labels (to be sure we would need to see what Inspector/Data is showing).

Try giving it a shot:

<img width="870" alt="Screenshot 2023-11-08 at 11 08 22" src="https://github.com/grafana/grafana/assets/745532/8041fefe-7a3e-4a43-85d7-e61c8846a4bc">

For more context. Here's how Loki returns data with and without `line_format`

| line_format | no line_format |
|---|---|
| <img width="1529" alt="Screenshot 2023-11-08 at 11 09 21" src="https://github.com/grafana/grafana/assets/745532/00b20a49-a283-48f0-9576-43e51b8b13aa"> | <img width="1535" alt="Screenshot 2023-11-08 at 11 09 30" src="https://github.com/grafana/grafana/assets/745532/082cdbcf-3193-4ef1-9a1d-d90bb9cfc85c"> |

In the table you can see all fields that can be used as variables or have transformations applied to them. When `line_format` is applied `Line` field changes and the original JSON is gone (so transformation doesn't work any more). But if trace_id is stored in `labels` you should be able to get it. It seems like you have it in labels, at least it's what I see in video you posted so it should work!

<img width="373" alt="Screenshot 2023-11-08 at 11 12 11" src="https://github.com/grafana/grafana/assets/745532/ea4fbdad-20f6-4e5a-9cc7-bc1955e0b422">

here `${labels}` variable contains entire JSON and you can apply regex transformation on it.

**Comment by JakeCooper** [comment]
I do indeed have a derived field called trace_id. It doesn't seem to work otherwise?

No dice with the above settings:

![CleanShot 2023-11-08 at 13 28 36](https://github.com/grafana/grafana/assets/5499880/05ec9960-48fc-4972-a69a-21e6e7053b75)

I mean it works if you have the trace_id formatted in the logfmt but not without

With:
![CleanShot 2023-11-08 at 13 29 11](https://github.com/grafana/grafana/assets/5499880/c15c744f-c90b-4662-9c79-99acc061e1c0)

Without:
![CleanShot 2023-11-08 at 13 28 32](https://github.com/grafana/grafana/assets/5499880/8ea1e93a-acef-42b4-8876-3c915dcad936)

**Comment by ifrost** [comment]
The tricky bit is that you need to have your expression without space before the colon `"trace_id":"(\w+)"` when applied on `labels` field. It's hard to figure out why (more details below). The new editor with transformations support (available in main and to be released in 10.3.0) should make it much smoother.

Here's what happens:
* The link is shown only if a variable `trace_id` exists. The variable is created for each field in the response (field names matches the name of the variable). Transformations allow adding additional variables.
* When you define `trace_id` derive field in Loki config it always creates a new field with that name (even if regex in derived field doesn't match anything - hence sometimes it shows as `null`). But it also means that variable `trace_id` is always created. It's just a field (column in the response data), the link is not created yet, but it means `trace_id` variable always exists.
* Now, when you define a transformation using regex and map it to the same name (`trace_id`) - it will override existing variable **only** if the regex matches anything
* In your examples transformation doesn't match anything but the link is created because it picks up `trace_id` from derived fields
  * after applying line_format, `Line` field doesn't contain trace_id and transformation doesn't extract anything, `trace_id` is not overriden
  * applying regex on `labels` requires skipping all spaces (because JSON in `labels` is stringified before transfomration is applied). In your case there's a space so regex doesn't match anything
  * in both cases the button shows up only because there's a derived field that provides `trace_id` variable (but is null). If you remove derived field the link won't show up at all (no field `trace_id` + no matching regex transform = no variable `trace_id`)

To recap: Try apply regex on `labels` field but make sure no spaces are used between `"` and `:` in the expression, e.g. `"trace_id":"(\w+)"`. This is because in the payload `labels` is an Object and is stringified without formatting. This should work regardless of applying `line_format` in the query.

**Comment by JakeCooper** [comment]
Unfortunately the above doesn't work either :(

![CleanShot 2023-11-09 at 15 24 25](https://github.com/grafana/grafana/assets/5499880/d3c949b2-7641-4f3b-b7b7-84d43b0dd239)

I think I sorta understand the mechanics. Will try and fiddle. If I can't get it, all good, but wondering when 10.3 is slated

**Comment by JakeCooper** [comment]
Oh shit wait I got it! I had a comma in there as well but I guess it was the last item in the json so no comma

Removing it works!

![CleanShot 2023-11-09 at 15 34 28](https://github.com/grafana/grafana/assets/5499880/9445afc7-4061-4348-8b3c-2294e3537be1)


![CleanShot 2023-11-09 at 15 33 12](https://github.com/grafana/grafana/assets/5499880/c712cefe-0982-4f9d-91ec-34877ccecca1)
